text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,335 @@
+
+For example, we could propose a feature called `force-tiny-fmt`, which would use different algorithms to implement `fmt` for use on resource constrained systems. The developer of the root crate would be able to choose the default behavior, or the `force-tiny-fmt` behavior while still retaining the ability of using a stable compiler.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A reference-level explanation is made for each of the items enumerated above.
+
+## 1 - Allow developers of root crates to recompile `core`
+
+### Use Case
+
+For developers working with new targets not yet supported by the Rust project, this feature would allow the compilation of `core` for any target that can be specified as a valid [custom target specification].
+
+[custom target specification]: https://rust-lang.github.io/rfcs/0131-target-specification.html
+
+This functionality would be possible even with the use of a stable compiler.
+
+Users of a nightly compiler would be able to set compile time feature flags for `core` through settings made in their `Cargo.toml`.
+
+### Caveats
+
+For users of a stable compiler, it would not be possible to modify the source code contents of `core`, or change any compile time features of `core` from the defaults used when publishing pre-compiled versions of `core`.
+
+The source code used to build `core` would be the same as the compiler used for building the current project.
+
+### User Interaction
+
+When compiling for a non-standard target, users may specify their target using a target specification file, rather than a pre-defined target.
+
+> NOTE: The current target specification is described in JSON, and contains some
+> implementation details regarding the use of LLVM as the compiler backend. This
+> RFC does not prescribe any changes to the Target Specification format, and is
+> intended to work with whatever the current/stable method of specifying a
+> custom target is.
+
+For example, currently a user may cross-compile by specifying a target known by Rust:
+
+```sh
+cargo build --target thumbv7em-none-eabihf
+```
+
+Users would also be able to specify a target specification file, by providing a path to the file to be used.
+
+```sh
+cargo build --target thumbv7em-freertos-eabihf.json
+```
+
+In general, any of the following would prompt Cargo to recompile `core`, rather than use a pre-compiled version:
+
+* A custom target specification is used
+* The root crate has modified the feature flags of `core`
+* The root crate has set certain profile settings, such as opt-level, etc.
+* The root crate has specified a `patch.sysroot` (this is defined in a later section)
+
+Users of a stable compiler would not be able to customize `core` outside of these profile settings.
+
+For users of a nightly compiler, compile time features of `core` may be specified using the same syntax used for other crate dependencies. These specified features may include unstable features.
```

Comment:
@SimonSapin the problem is that which libraries are special-cased to the implementation is necessarily an implementation specific detail. We want to move creates to crates.io when they become 100% stable code and not have things break. I think the compiler should provide a "workspace override" which does the `[patch.sysroot]`-ing and/or we start to distinguish the "default source" from `crates.io` (the default source may be a composition of sources).

Reply:
@Ericson2314 I’m sorry. I feel like I should respond to your message here, but I have a very hard time parsing it :/ It seems based on multiple unstated assumptions. For example:

> an implementation specific detail

Are you imagining a world where there are multiple implementations of Rust where they each have their take on the entire toolchain? Today mrustc exists, but doesn’t have its own Cargo.

My objection was to hard-coding a list of names known to be crates of the standard library *in Cargo*, because Cargo and the standard library are developed separately.

> We want to move creates to crates.io when they become 100% stable code and not have things break.

I know from previous interactions with you that you have this idea of the standard library  somehow having its source of truth on crates.io. You take it as a given here but I still maintain that this idea is fundamentally contradictory, because of the difference in versioning between these two worlds.

A library gets to pick what other libraries from crates.io it depends on, and it gets to pick their version numbers. A program can even end up with multiple versions of the “same” library in its dependency graph.

A library does *not* however get to pick what version of the compiler is being used. (Only which versions it tries to be compatible with.) The standard library is by definition what ships with a given compiler. It doesn’t have its own version number, it shares the compiler’s version number.

I think that one can make a valid argument that less functionality should be the responsibility of the standard library. For example, `std::sync::mpsc` could be deprecated without a replacement in `std` and https://crates.io/crates/crossbeam-channel be recommended instead. But being shipped with the compiler is what “standard library” means. “Moving `std` to crates.io” doesn’t make any sense in my opinion.

Reply:
That's how the standard library works today, but there's no rule that it must always be that way.  It's true that some parts of it, mostly in `core`, are tightly bound to the compiler implementation, such as wrappers around intrinsics and definitions of language items.  But significant chunks are not, including most of `std` itself (all of the wrappers around OS functionality) as well as significant parts of `core` (e.g. unicode, flt2dec, iterator adapters).  The current implementations of those APIs may or may not use unstable features, but the APIs don't inherently require it.  The deprecation route you mentioned could work in theory, but I don't think users would appreciate a large percentage of the standard library being deprecated.

Personally, I would like to see a new "even more core than `core`" crate that has the absolute bare minimum amount of code required to expose all of the compiler's functionality.  Then the rest can be moved to `crates.io`, and can be made optional – useful for cases where, say, you're extremely constrained on binary size and want to be in control of every single line of code that actually makes it into the output binary.

Reply:
> Then the rest can be moved to crates.io

`rand` moved from the standard library to crates.io. Because it was not marked at stable in `std` (or it was before 1.0, I don’t remember), we could simply remove it from there.

But there’s not a lot of that left that we can do. For APIs that are `#[stable]` in `std`, we promised to keep having them in `std`. At most when can deprecate them, which as you noted has downsides. With that in mind, what does “move to crates.io” even mean? Something can not move if it also stays at its current location. Are you saying you want two copies of the same APIs? How is that useful? Can they diverge, in particular if the crates.io one wants to make breaking changes in a 2.0 version?

> can be made optional

This RFC proposes cargo feature flags for standard library crates, which presumably will allow making parts of the standard library optional: you can disable them if you don’t use them. How does “move to crates.io” help with *not* using something?

> for cases where, say, you're extremely constrained on binary size

The decision to move Unicode tables into libcore was based explicitly on them not affecting the binary size of programs that don’t use them, with confirmation from the Embedded WG. (The linker eliminates unused symbols.) So I feel that binary size is not a valid argument for a smaller standard library.

Reply:
@SimonSapin 

> Are you imagining a world where there are multiple implementations of Rust where they each have their take on the entire toolchain? Today mrustc exists, but doesn’t have its own Cargo.
>
> My objection was to hard-coding a list of names known to be crates of the standard library in Cargo, because Cargo and the standard library are developed separately.

You may be surprised that I agree with that. When I said implementation-specific I meant *rustc*, not Cargo. Much unstable exists because it is more tightly coupled with rustc. A different compiler or interpreter may need to be tightly coupled in a different way.

> You take it as a given here but I still maintain that this idea is fundamentally contradictory, because of the difference in versioning between these two worlds. [...]

@SimonSapin I hear what you are saying. But I don't see any contradictions. To be clear, we agree that `std` needs to continue to export everything it does today. \[Maybe some future additions can allow opt-out of that, but let's not imagine such things. I don't want my arguments to hinge on that.] But nothing says it cannot *re*-export stuff rather than implement it itself.

Maybe we don't agree that reexporting like this is *desirable*, but let's first agree that it's *possible*. OK?

Reply:
@comex I don't think we can have a *single* compiler-specific crate unless we allow crates to have cyclic dependencies, but yes glad to hear we both want a clean separation between rustc-specific and rustc-agnostic code.

Reply:
> The decision to move Unicode tables into libcore was based explicitly on them not affecting the binary size of programs that don’t use them, with confirmation from the Embedded WG. (The linker eliminates unused symbols.) So I feel that binary size is not a valid argument for a smaller standard library.

I know.  But if you don't want something in the binary, it's cleaner to have it not present at all, rather than needing to manually avoid calling something that's always in scope (e.g. `char::is_whitespace`).  This is especially weird with the format machinery (which is baked into libcore functions like `panic_bounds_check`); I see [a comment of yours from last year](https://github.com/rust-lang/rfcs/pull/2305#issuecomment-408264884) saying it is possible to avoid it being linked in, but it's definitely nonintuitive what exactly the constraints are.

But sure, a minimal core could also be achieved by putting everything else behind Cargo feature flags, rather than actually creating a separate crate.

By the way, another use case for a minimal core is if you just want to write your own standard library with a different design, and want to be able to replace things like impls on primitives (`impl char`, `impl<T> [T]`, `impl<T> *const T`), or the precise list of methods on `Iterator` (even if some basic functionality has to be fixed for the compiler to codegen `for` loops).  Those things are lang items, so letting third-party crates implement them on stable would require additional work.  However, a minimal core is a prerequisite.

> With that in mind, what does “move to crates.io” even mean? Something can not move if it also stays at its current location. Are you saying you want two copies of the same APIs? How is that useful? Can they diverge, in particular if the crates.io one wants to make breaking changes in a 2.0 version?

As I imagine it, the source of truth would be on crates.io, but the Rust distribution would also include a copy mainly for backwards compatibility purposes.  A 2.0 could be done in theory but is probably inadvisable.  Benefits of having a separate, stable-source crate include:
- Ability to patch the source on stable (this RFC already proposes `patch` functionality but only for nightly).
- Ability for conservatively-minded users to update to a newer compiler while holding back on changes to `std`, e.g. the recent wholesale replacement of `HashMap` with `hashbrown`  or the upcoming replacement of synchronization primitives with `parking_lot`.
- On the flipside: ability for experimentally-minded users to test such changes in advance of a full release, without having to also change their compiler configuration (which makes comparison more difficult).